dependencies/README.md

```diff
@@ -88,3 +90,89 @@ The tool allows users to choose the scope of entry point classes:
   - **All classes in the input class path**: when reachability check is off, then
     all static linkage errors from all classes in the classpath, regardless of the reachability,
     are reported.
+
+<a name="maven_dependency_graph"></a>
+# Maven Dependency Graph
+
+A Maven dependency graph is a graph data structure where
+- Node: a node is a Maven artifact identified by [Maven coordinates][1] such as
+  `com.example.foo:bar:1.5.4`. A node may be marked _unavailable_ when the artifact is not available
+  in Maven repositories.
+- Edge: a directed edge is a dependency from one Maven artifact (_source_ of the edge)
+  to another Maven artifact (_target_ of the edge).
+
+  A dependency has a boolean attribute `optional` and an enum attribute `scope`,
+  among other properties listed in [POM Reference: Dependencies][2].
+
+  Self-loops are not possible. A parallel edge is allowed but is dropped in the model.
+
+## Comparison with Maven Dependency Trees by pom.xml
+
+Compared to a _Maven dependency tree_ retrieved by a pom.xml through
+`RepositorySystem.resolveDependencies`, a Maven dependency graph has the following differences:
+
+- A Maven dependency tree cannot have multiple versions of a Maven artifact, whereas a Maven
+  dependency graph can have them.
+- A Maven dependency tree does not have transitive `scope: provided` dependencies under
+  a `scope: provided` dependency, whereas a Maven dependency graph can have them.
+- A Maven dependency tree does not contain a cycle,
+  whereas a Maven dependency graph can contain cycles.
+
+## Graph Construction
+
+Given an ordered list of Maven artifacts, `DependencyGraphBuilder` constructs a Maven dependency
+graph in the following steps:
+
+1. Start with a graph with nodes of the Maven artifacts in the list.
+   The nodes are called _initial nodes_.
+2. Pick up an unvisited node which is not marked _unavailable_ in a graph in breadth-first manner.
+3. By reading dependencies of the node,
+   - add new nodes corresponding to the target Maven artifacts, identified by Maven coordinates,
+     if not present.
+   - add edges from the source node to the target nodes of the Maven artifacts.
+4. Repeat Step 2-3, until all nodes are visited in the breadth-first traversal.
+
+## Graph Construction Failure
+
+A graph construction may _fail_ when there is a problem.
+
+### Unavailable Artifact
+
+A Maven artifact may be unavailable through Maven repositories. Such nodes are marked unavailable
+at Step 3 of a graph construction.
+
+An unavailability of Maven artifact is called _safe_ when the path from the initial nodes to
```

Comment:
I'm not sure "safe" gives the right impression here. "Unnecessary"? Optional would be better but it's taken. Maybe just optional or provided?

Reply:
Updated to use 'optional' and 'provided'